lib/src/api_endpoints.dart

```diff
@@ -0,0 +1,125 @@
+import 'constants.dart';
+import 'sdk_key.dart';
+
+/// Utility class for constructing Eppo API base URLs.
+///
+/// Determines the effective base URL considering baseUrl, subdomain from SDK token,
+/// and defaultUrl in that order.
+class ApiEndpoints {
+  /// The SDK key
+  final SDKKey sdkKey;
+  
+  /// Custom base URL (optional)
+  final String? baseUrl;
+  
+  /// Default base URL
+  final String defaultUrl;
+
+  /// Creates a new ApiEndpoints instance.
+  ///
+  /// [sdkKey] SDK Key instance for subdomain
+  /// [baseUrl] Custom base URL (optional)
+  /// [defaultUrl] Default base URL (defaults to precomputedBaseUrl)
+  ApiEndpoints._({
+    required this.sdkKey,
+    this.baseUrl,
+    required this.defaultUrl,
+  });
+
+  /// Creates a new ApiEndpoints instance configured for precomputed flags.
+  ///
+  /// [sdkKey] SDK Key instance for subdomain
+  /// [baseUrl] Custom base URL (optional)
+  static ApiEndpoints precomputed({
+    required SDKKey sdkKey,
+    String? baseUrl,
+  }) {
+    return ApiEndpoints._(
+      sdkKey: sdkKey,
+      baseUrl: baseUrl,
+      defaultUrl: precomputedBaseUrl,
+    );
+  }
+
+  /// Gets the normalized base URL based on the following priority:
+  /// 1. If baseUrl is provided and not equal to the default URL, use it
+  /// 2. If the SDK Key contains a subdomain, prepend to the defaultUrl
+  /// 3. Otherwise, fall back to DEFAULT_BASE_URL
+  ///
+  /// The returned URL will:
+  /// - Always have a protocol (defaults to https:// if none provided)
+  /// - Never end with a trailing slash
+  ///
+  /// @return The normalized base URL
+  String getBaseUrl() {
+    String effectiveUrl;
+
+    if (baseUrl != null && baseUrl != defaultUrl) {
+      // This is to prevent forcing the SDK to send requests to the CDN server without
+      // a subdomain even when encoded in SDK Key.
+      effectiveUrl = baseUrl!;
+    } else if (sdkKey.isValid()) {
+      final subdomain = sdkKey.getSubdomain();
+      if (subdomain != null) {
+        // Extract the domain part without protocol
+        final domainPart = defaultUrl.replaceAll(RegExp(r'^(https?:)?//'), '');
+        effectiveUrl = '$subdomain.$domainPart';
+      } else {
+        effectiveUrl = defaultUrl;
+      }
+    } else {
+      effectiveUrl = defaultUrl;
+    }
+
+    // Remove any trailing slashes
+    effectiveUrl = effectiveUrl.replaceAll(RegExp(r'/+$'), '');
+
+    // Add protocol if missing
+    if (!effectiveUrl.contains(RegExp(r'^(https?://|//)'))) {
+      effectiveUrl = 'https://$effectiveUrl';
+    }
+
+    return effectiveUrl;
+  }
+
+  /// Gets the URL for the precomputed flags endpoint.
+  ///
+  /// @param queryParams Optional query parameters to append to the URL
+  /// @return The full precomputed flags endpoint URL
+  String getPrecomputedFlagsEndpoint([Map<String, String>? queryParams]) {
+    final baseEndpoint = '${getBaseUrl()}$precomputedFlagsEndpoint';
+    
+    if (queryParams == null || queryParams.isEmpty) {
```

Comment:
```suggestion
    if (queryParams?.isEmpty) {
```

Reply:
Dart doesn't coerce null into a boolean `false` so dart complains if I use the nullable accessor.
I'm not sure the alternative is much of an improvement here.

```dart
    if (queryParams?.isEmpty ?? false) {
      return baseEndpoint;
    }
```